models/staging/ga4/stg_ga4__user_id_mapping.sql

```diff
@@ -0,0 +1,29 @@
+with events_with_user_id as (
+    select 
+        user_id,
+        user_pseudo_id,
+        event_timestamp 
+    from {{ref('stg_ga4__events')}}
+    where user_id is not null
+        and user_pseudo_id is not null
+),
+include_last_seen_timestamp as (
+    select 
+        user_id,
+        user_pseudo_id,
+        max(event_timestamp) as last_seen_timestamp
```

Comment:
If someone arrives at a site not logged in or cookied, they start the session with a `user_pseudo_id` but no `user_id`. They complete a few page_view events and then log in setting a `user_id`.  The navigate the logged in portion registering a few page_view events before logging out. Subsequent page_views will not have a `user_pseudo_id`. This is how the GA4 interface applies user IDs, but Google has always said that if you want to stitch this data together within sessions and across sessions (which is a situation that doesn't apply to this comment), then you should do that in BQ.

I'm willing to approve this PR as an improvement over what we have but I wouldn't call the `user_id` problems resolved with this PR. I think we need to get the maximum timestamp by `user_pseudo_id` and then attach that to the most recent, non-null `user_id` to completely resolve `user_id` issues within a session.

Reply:
As far as I can tell, what's implemented is what you described in your last paragraph. This model looks for the most recent, non-null `user_id` for each `user_pseudo_id`

Are you saying this should be done session-by-session rather than across all events?

Reply:
Yes, you are right that it maps the last `user_id` to the `user_pseudo_id`. See my comment further down, though.